blocks/algorithms/__init__.py

```diff
@@ -0,0 +1,239 @@
+        `cost` computation graph will be considered parameters.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+        Some shared variables are not parameters (e.g. those created by
+        random streams)
+
+    .. todo::
+
+        Due to a rather premature status of the :class:`ComputationGraph`
+        class the parameter used only inside scans are not fetched
+        currently.
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, cost, params=None):
+        self.cost = cost
+        self.params = (params if params
+                       else ComputationGraph(cost).get_shared_variables())
+        self._cost_computation_graph = ComputationGraph(self.cost)
+        self._updates = []
+
+    @property
+    def inputs(self):
+        """Return inputs of the cost computation graph.
+
+        Returns
+        -------
+            list of Theano variables
+
+        """
+        return self._cost_computation_graph.inputs
+
+    @property
+    def updates(self):
+        return self._updates
+
+    @updates.setter
+    def updates(self, value):
+        self._updates = value
+
+    def add_updates(self, updates):
+        """Add updates to the training process.
+
+        The updates will be done _before_ the parameters are changed.
+
+        Parameters
+        ----------
+        updates : list of tuples or :class:`OrderedDict`
+            The updates to add.
+
+        """
+        if isinstance(updates, OrderedDict):
+            updates = list(updates.items())
+        assert isinstance(updates, list)
+        self.updates.extend(updates)
+
+
+class GradientDescent(DifferentiableCostMinimizer):
+    """A base class for all gradient descent algorithms.
+
+    By "gradient descent" we mean a training algorithm of the following
+    form:
+
+    .. code-block::  python
+
+        for batch in data:
+            for param in params:
+                param += step.compute(grad_wr_param_on_batch)
+
+    Parameters
+    ----------
+    step_rule : instance of :class:`StepRule`
+        An object incapsulating most of the algorithm's logic. Its
+        `compute_step` method is called to get a Theano expression
+        for the actual step to take for each parameter. Note, that
+        the step rule might have a state, e.g. to remember a weighed
+        sum of gradients from previous steps like it is done in
+        gradient descent with momentum. If ``None``, an instance of
```

Comment:
Nitpicking, but instead of saying "If `None`" (which suggests that the user is expected to pass that) you could say
```
step_rule : instance of :class:`StepRule`, optional
    Does this and that. By default, ...
```